### DISCUSSION TITLE
Thoughts on Spring Cloud AWS 3.0

### DISCUSSION THREAD

**Discussion by eddumelendez** [awspring/spring-cloud-aws#63] [Ideas]
Spring Cloud AWS 3.0 has been discussed the last year with great topics in mind. However, I see some improvements should be done in the next minor version `2.4` instead of start working on version 3. Some of this improvements are:

* Improve and provide a S3 module
* Parameter Store and Secrets Manager starters should move classes to `autoconfigure` module
* Drop config from Parameter Store and Secrets Manager starters
* Rationalize about RDS integration
* Rationalize about ElastiCache integration
* Deprecate `context` module

I hope we can integrate [appconfig](https://github.com/spring-cloud/spring-cloud-aws/pull/652) and have a [PR for cloudmap integration](https://github.com/awspring/spring-cloud-aws/issues/5) for 2.4 but let's see if it can be made

Why?

* There are still some topics to review regarding to move the project to aws sdk v2 as we can see [here](https://github.com/spring-cloud/spring-cloud-aws/pull/524). We don't want to start migrating and see that it will take more time that we really need for a successful delivery. This can be done in parallel with 2.4 tho but it will not be the official branch yet.
* 2.4 will allow us to improve the project with the topics mentioned above having in mind backward compatibility. So, when 3.0 arrives we can remove deprecated classes, modules and start with a desired state.

Also, Spring Boot 3 will start using java 11 and I think it is a great opportunity to meet with Spring Cloud AWS 3. Although, this should not be true I think it make sense.

@maciejwalkowiak @MatejNedic wdyt?

**Comment by maciejwalkowiak** [comment]
I am not sure it makes sense to put efforts into development code base that is doomed to be deprecated relatively soon. Considering time constraints I would prefer to keep 2.x for minor changes - we can still release 2.4 especially if there will be community contributions, but to put efforts into 3.x. 

There is quite a bit to learn about how to use AWS SDK v2 but I think if we split the work properly, it is doable to be ready with 3.0 in Q2 2021. If we continue working on 2.x, I am afraid it will move the 3.0 to the end of the year.

**Reply by eddumelendez** [reply]
IMO, spring cloud aws 3, given it is a major version should contain:
- spring boot 3
- spring cloud compatible with spring boot 3
- run with java 11
- aws sdk v2

The current approach which I think has been taking is:
- spring boot 2.4. maybe 2.5 compatible?
- run with java 8
- aws sdk v2

So, if it is as was described, it means a minor version 3.1 will raise a major change:
- spring boot 3
- spring cloud compatible with spring boot 3
- run with java 11

**Reply by maciejwalkowiak** [reply]
I prefer not having a dependency on the Spring Boot 3.0 release. We can make Spring Cloud AWS 3.0 compatible with Spring Boot 2.x and Java 11. Then we can release it as soon as we are ready.

Spring Boot 3.0 and Java 11 version can be a bump to Spring Cloud AWS 4.0

**Reply by sshevlyagin** [reply]
I'm really curious, any more thoughts on supporting AWS SDK v2 and the timing?

**Reply by choweiyuan** [reply]
Any updates on the timeline? Still aiming for end of year?

**Reply by maciejwalkowiak** [reply]
We're closing now 2.3.x and soon after release 2.4. Then we get back to 3.0 but this time without giving vague promises. We are going to release milestones, so first release will contain just a subset of integrations. Stay tuned and apologies for poor communication.

**Comment by maciejwalkowiak** [comment]
Spring Boot 3 will use Java 17 as a baseline - it will be released in Q4 AFAIK and Spring Cloud will follow probably in Q1 2023. Ideally, we are faster than that. So I prefer to stick to Java 8 for now and if it takes longer than expected, bump Java version and take advantage of Java 17 at the later stage.

**Comment by maciejwalkowiak** [comment]
Another thing we must have in 3.0 - all public APIs must be annotated with `@Nullable`/`@NonNull`, either from `org.springframework.lang` or `org.jetbrains.annotations`. This will improve Kotlin compatibility and also makes the API easier to use. 

I am leaning towards `org.jetbrains.annotations` because it has more useful annotations like marking API as experimental.

**Comment by maciejwalkowiak** [comment]
Considering current state, we could release M1 with:
- core - to configure credentials and region
- Parameter Store
- Secrets Manager

Both integrations have to be updated to reflect changes we've made in 2.3.3 and 2.3.4.
